File: c/complex-numbers/src/complex_numbers.c
#include "complex_numbers.h"

complex_t c_add(complex_t a, complex_t b)
{
        complex_t r = {a.real + b.real, a.imag + b.imag}; 

        return r; 
}

complex_t c_sub(complex_t a, complex_t b)
{
        complex_t r = {a.real - b.real, a.imag - b.imag}; 

        return r; 
}

complex_t c_mul(complex_t a, complex_t b)
{
        complex_t r = {a.real * b.real - a.imag * b.imag, 
                       a.imag * b.real + a.real * b.imag}; 

        return r; 
}

complex_t c_div(complex_t a, complex_t b)
{
        complex_t r; 

        double denom = (b.real * b.real + b.imag * b.imag);

        r.real = (a.real * b.real + a.imag * b.imag) / denom; 
        r.imag = (a.imag * b.real - a.real * b.imag) / denom; 

        return r; 
}

double c_abs(complex_t x)
{
        double r = sqrt(x.real * x.real + x.imag * x.imag); 

        return r; 
}

complex_t c_conjugate(complex_t x)
{
        complex_t r = {x.real, 0 - x.imag}; 
        return r; 
}

double c_real(complex_t x)
{
        return x.real; 
}

double c_imag(complex_t x)
{
        return x.imag; 
}

complex_t c_exp(complex_t x)
{
        complex_t euler_id = {cos(x.imag), sin(x.imag)}; 

        double a = exp(x.real); 

        complex_t r = {a * euler_id.real, a * euler_id.imag}; 

        return r; 
}

File: python/complex-numbers/complex_numbers.py
from math import sqrt, exp, cos, sin

class ComplexNumber(object):
    def __init__(self, real, imaginary):
        self.real = real
        self.imaginary = imaginary

    def __eq__(self, other):
        return self.real == other.real and self.imaginary == other.imaginary

    def __add__(self, other):
        r = self.real + other.real
        i = self.imaginary + other.imaginary

        return ComplexNumber(r, i)

    def __mul__(self, other):
        r = self.real * other.real - self.imaginary * other.imaginary
        i = self.imaginary * other.real + self.real * other.imaginary

        return ComplexNumber(r, i)

    def __sub__(self, other):
        r = self.real - other.real
        i = self.imaginary - other.imaginary

        return ComplexNumber(r, i)

    def __truediv__(self, other):
        a = self.real
        b = self.imaginary
        c = other.real
        d = other.imaginary

        r = (a * c + b * d) / (c * c + d * d)
        i = (b * c - a * d) / (c * c + d * d)

        return ComplexNumber(r, i)

    def __abs__(self):
        return sqrt(self.real**2 + self.imaginary**2)

    def conjugate(self):
        r = self.real
        i = -self.imaginary

        return ComplexNumber(r, i)

    def exp(self):
        ea = exp(self.real)

        eb = ComplexNumber(cos(self.imaginary), sin(self.imaginary))

        return ComplexNumber(ea, 0) * eb
